File: backend/vm/journal.py
"""VM Mutation types — the only output the VM produces.

All dataclasses are frozen (immutable). The VM produces a Journal (list of
Mutation) and the engine replays it against mutable battle state.
"""

from dataclasses import dataclass, field
from typing import Union


@dataclass(frozen=True)
class StatChange:
    """Permanent stat modification applied to a sprite."""
    target: str       # "sprite_self" | "sprite_opp"
    stat: str         # "atk" | "def" | "sp_atk" | "sp_def" | "speed" | ...
    steps: int        # positive = buff, negative = debuff
    scope: str = "battlefield"  # "battlefield" | "persistent" | "permanent"
    # Optional metadata for engine-side resolution
    source: str | None = None         # trait/skill name for replace-mode clearing
    element: str | None = None        # element filter ("火", "each", etc.)
    per_element: int | None = None    # max per element when element="each"
    on_next: bool = False             # defer to next matching skill use
    if_type: str | None = None        # "attack" | "defense" | "status"
    skill_filter: str | None = None   # "attack" | "defense" | "status" | "all" | "others" | "adjacent" | "bare_*"
    skill_where: dict | None = None   # per-skill conditional {"q": ..., "op": ..., "value": ...}


@dataclass(frozen=True)
class ModifierInjection:
    """Internal modifier for the skill pipeline (not applied to sprite state).

    Engine collects these during power/mult phases and feeds them into
    the damage formula. Examples: power_mult, damage_mult, damage_reduction,
    energy_cost modifications.
    """
    target: str       # "skill_off_0" | "sprite_self" | ...
    stat: str         # "power_mult" | "damage_mult" | "damage_reduction" | "energy_cost" | ...
    value: float
    scope: str = "battlefield"  # "battlefield" | "persistent" | "permanent"
    mode: str = "set" # "set" | "add" | "multiply"
    # Optional metadata for engine-side resolution
    name: str | None = None           # exact skill name or devotion name
    source: str | None = None         # trait/skill name for replace-mode clearing
    element: str | None = None        # element filter ("火", "each", etc.)
    per_element: int | None = None    # max per element when element="each"
    on_next: bool = False             # defer to next matching skill use
    if_type: str | None = None        # "attack" | "defense" | "status"
    skill_filter: str | None = None   # "attack" | "defense" | "status" | "all" | "others" | "adjacent" | "bare_*"
    skill_where: dict | None = None   # per-skill conditional {"q": ..., "op": ..., "value": ...}
    ttl: int = 0                      # remaining turns (0 = no expiry)
    then: list | None = None          # chained effects (devotion then-block)


@dataclass(frozen=True)
class Damage:
    """Final computed damage to apply to a sprite."""
    target: str       # "sprite_opp" | "sprite_self"
    amount: int
    element: str
    type: str         # "物攻" | "魔攻"


@dataclass(frozen=True)
class Heal:
    """HP restoration."""
    target: str
    amount: int


@dataclass(frozen=True)
class EnergyChange:
    """Energy gain or loss."""
    target: str
    delta: int        # positive = gain, negative = lose


@dataclass(frozen=True)
class MarkChange:
    """Mark stack change on a team."""
    target_team: str  # "own" | "opp"
    name: str         # mark name
    delta: int        # positive = add stacks, negative = remove
    action: str = "apply"   # "apply" | "dispel" | "steal" | "convert"
    ratio: float = 1.0      # convert: abnormal→mark conversion ratio
    source_abnormal: str | None = None  # convert: source abnormal name


@dataclass(frozen=True)
class AbnormalChange:
    """Abnormal status stack change on a sprite."""
    target: str       # "sprite_self" | "sprite_opp"
    name: str         # abnormal name, e.g. "中毒"
    delta: int        # stacks to add (positive) or set-to-zero (special sentinel)
    scope: str = "battlefield"  # "battlefield" | "persistent" | "permanent"


@dataclass(frozen=True)
class WeatherSet:
    """Weather change."""
    weather: str
    turns: int


@dataclass(frozen=True)
class Dispel:
    """Remove effects from a target."""
    target: str
    what: str         # "positive" | "negative" | "abnormal" | "mark"
    name: str | None = None
    limit: int | None = None
    type_limit: int | None = None
    source: str | None = None  # only dispel effects from this source


@dataclass(frozen=True)
class Steal:
    """Steal effects/energy from a target."""
    from_target: str  # "sprite_opp" | "team_opp"
    what: str         # "positive" | "mark" | "energy"
    name: str | None = None
    amount: int | None = None


@dataclass(frozen=True)
class Tick:
    """Trigger abnormal tick damage on a target."""
    target: str
    abnormal_name: str


@dataclass(frozen=True)
class Double:
    """Double a mark/effect on a target."""
    target: str
    what: str
    name: str | None = None


@dataclass(frozen=True)
class EffectDelta:
    """Add delta to all matching effects (positive/negative) on a target."""
    target: str       # "sprite_self" | "sprite_opp"
    what: str         # "positive" | "negative"
    delta: int        # stacks/steps to add


@dataclass(frozen=True)
class Charge:
    """Enter charging state."""
    target: str       # "sprite_self"


@dataclass(frozen=True)
class Escape:
    """Remove self from field (switch/escape)."""
    target: str
    inherit: bool = False
    urgent: bool = False
    then: list | None = None  # Effects to execute after escape


@dataclass(frozen=True)
class Return:
    """Return to field after escape/switch."""
    target: str


@dataclass(frozen=True)
class Lock:
    """Lock opponent from switching."""
    target: str
    turns: int


@dataclass(frozen=True)
class Interrupt:
    """Interrupt the target's current action."""
    target: str


@dataclass(frozen=True)
class Exchange:
    """Exchange something between sprites (HP ratio, effects, skills)."""
    target: str
    what: str         # "hp_ratio" | "effects" | "skills" | "adjacent_skills"


@dataclass(frozen=True)
class Reset:
    """Reset a stat to base value."""
    target: str
    stat: str


@dataclass(frozen=True)
class Redirect:
    """Redirect the next action to a different target."""
    target: str


@dataclass(frozen=True)
class BurstGrant:
    """Grant burst effects to matching skills."""
    target: str                         # "sprite_self"
    skill_where: dict | None = None     # per-skill conditional filter
    skill_filter: str | None = None     # "attack" | "defense" | "status" | "all"
    effects: tuple = ()                 # burst effect dicts to execute on first_action
    source: str = ""                    # trait name


@dataclass(frozen=True)
class Replay:
    """Replay a previously used skill or burst skill."""
    from_: str        # "sprite_self" | "team_burst" (from_ avoids Python keyword)
    skill_filter: dict | None = None


@dataclass(frozen=True)
class Borrow:
    """Borrow properties from the opponent's current skill."""
    from_skill: str   # "skill_opp_current"


@dataclass(frozen=True)
class TeamCounterDelta:
    """Write to a team-level counter."""
    target: str        # "own" | "opp"
    key: str           # counter key name
    delta: int         # +1 or -1


@dataclass(frozen=True)
class LivesDelta:
    """Modify player lives (魔力)."""
    target_team: str   # "own" | "opp"
    delta: int         # +1 (奉献) or -1 (魔力消耗)


@dataclass(frozen=True)
class ScheduleEntry:
    """Register delayed effects for a future turn (RISC: defer)."""
    turns: int
    at: str = "turn_start"    # "turn_start" | "turn_end"
    then: list = field(default_factory=list)  # IR effects to execute at the delayed time


@dataclass(frozen=True)
class InheritEffectsMutation:
    """Transfer effects between sprites on switch."""
    source_key: str             # "self" | "target"
    target_key: str             # "enemy_new" | sprite_ref
    scope: str = "battlefield"
    via_pending: bool = False   # route through battle.pending_effects
    effects: tuple = ()         # fixed effects to apply to incoming sprite
    inherit_stat_effects: bool = False  # copy dynamic stat effects


@dataclass(frozen=True)
class TransformMutation:
    """Transform a sprite's species and optionally skills."""
    species: str
    skills: tuple[str, ...] | None = None
    reset_hp: bool = False
    reset_energy: bool = False


@dataclass(frozen=True)
class TraitInteractionMutation:
    """Suppress, remove, or copy a trait."""
    action: str             # "suppress" | "remove" | "copy"
    target: str             # sprite ref
    copy_from: str | None = None
    new_ability: str | None = None


@dataclass(frozen=True)
class GainSkillsMutation:
    """Grant temporary skills to a sprite from a skill pool."""
    count: int = 1
    exclude_carried: bool = True
    source: str = "learnset"   # "learnset" | "global"
    target: str = "sprite_self"


@dataclass
class CounterRegister:
    """Register a persistent counter/watcher on the skill."""
    name: str | None = None
    cond: dict | None = None     # Trigger condition
    then: list = field(default_factory=list)  # IR effects to execute
    scope: str = "persistent"
    listen: frozenset | None = None  # explicit trigger set (None = auto-infer)
    threshold: int = 1            # fire every N triggers
    reset_on_fire: bool = True    # reset counter after firing


# Union of all mutation types the VM can produce
Mutation = Union[
    StatChange, ModifierInjection, Damage, Heal, EnergyChange,
    MarkChange, AbnormalChange, WeatherSet, Dispel, Steal, Tick,
    Double, EffectDelta, Charge, Escape, Return, Lock, Interrupt, Exchange,
    Reset, Redirect, Replay, Borrow, CounterRegister, BurstGrant,
    TeamCounterDelta, LivesDelta, ScheduleEntry,
    InheritEffectsMutation, TransformMutation, TraitInteractionMutation,
    GainSkillsMutation,
]

# Journal is an ordered list of mutations
Journal = list[Mutation]

File: native/roco-core/src/journal.rs
//! journal — VM 的 31 种 Mutation（移植自 backend/vm/journal.py）。
//!
//! VM 唯一的输出形态：一段不可变的 Mutation 序列，由引擎回放到
//! 可变战斗状态上。字段与 Python dataclass 一一对应；
//! 嵌套 IR 载荷（then/effects/cond/skill_where）暂用 serde_json::Value
//! 承载（executor 移植时决定是否升级为强类型）。

use crate::resolve::Val;
use serde::{Deserialize, Serialize};

/// 通用目标引用："sprite_self" | "sprite_opp" | "team_own" | "team_opp" | ...
pub type TargetRef = String;

/// 嵌套 IR 载荷占位（效果树 / 条件树 / 技能过滤条件）。
/// 语义由 executor 按编译器默认值规则解释。
pub type IrPayload = serde_json::Value;

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(tag = "op", rename_all = "snake_case")]
pub enum Mutation {
    /// 永久属性等级变化（StatChange）
    StatChange {
        target: TargetRef,
        stat: String,
        steps: i64,
        #[serde(default = "default_battlefield")]
        scope: String,
        #[serde(default)]
        source: Option<String>,
        #[serde(default)]
        element: Option<String>,
        #[serde(default)]
        per_element: Option<i64>,
        #[serde(default)]
        on_next: bool,
        #[serde(default)]
        if_type: Option<String>,
        #[serde(default)]
        skill_filter: Option<String>,
        #[serde(default)]
        skill_where: Option<IrPayload>,
    },
    /// 技能管线内部修饰符（ModifierInjection）。
    /// value 保留 Python 动态类型（flag_set 传 True/int 原样）。
    ModifierInjection {
        target: TargetRef,
        stat: String,
        #[serde(default)]
        value: Val,
        #[serde(default = "default_battlefield")]
        scope: String,
        #[serde(default = "default_set")]
        mode: String,
        #[serde(default)]
        name: Option<String>,
        #[serde(default)]
        source: Option<String>,
        #[serde(default)]
        element: Option<String>,
        #[serde(default)]
        per_element: Option<i64>,
        #[serde(default)]
        on_next: bool,
        #[serde(default)]
        if_type: Option<String>,
        #[serde(default)]
        skill_filter: Option<String>,
        #[serde(default)]
        skill_where: Option<IrPayload>,
        #[serde(default)]
        ttl: i64,
        #[serde(default)]
        then: Option<Vec<IrPayload>>,
    },
    /// 最终伤害（Damage）
    Damage {
        target: TargetRef,
        amount: i64,
        element: String,
        #[serde(rename = "type")]
        attack_type: String,
    },
    /// HP 恢复（Heal）
    Heal { target: TargetRef, amount: i64 },
    /// 能量增减（EnergyChange）
    EnergyChange { target: TargetRef, delta: i64 },
    /// 印记层数变化（MarkChange）
    MarkChange {
        target_team: String,
        name: String,
        delta: i64,
        #[serde(default = "default_apply")]
        action: String,
        #[serde(default = "default_one_f")]
        ratio: f64,
        #[serde(default)]
        source_abnormal: Option<String>,
    },
    /// 异常状态层数变化（AbnormalChange）
    AbnormalChange {
        target: TargetRef,
        name: String,
        delta: i64,
        #[serde(default = "default_battlefield")]
        scope: String,
    },
    /// 天气变更（WeatherSet）
    WeatherSet { weather: String, turns: i64 },
    /// 驱散（Dispel）
    Dispel {
        target: TargetRef,
        what: String,
        #[serde(default)]
        name: Option<String>,
        #[serde(default)]
        limit: Option<i64>,
        #[serde(default)]
        type_limit: Option<i64>,
        #[serde(default)]
        source: Option<String>,
    },
    /// 偷取（Steal）
    Steal {
        from_target: TargetRef,
        what: String,
        #[serde(default)]
        name: Option<String>,
        #[serde(default)]
        amount: Option<i64>,
        #[serde(default = "default_steal")]
        action: String,
    },
    /// 异常 tick 伤害（Tick）
    Tick { target: TargetRef, abnormal_name: String },
    /// 印记/效果翻倍（Double）
    Double {
        target: TargetRef,
        what: String,
        #[serde(default)]
        name: Option<String>,
    },
    /// 全体正/负效果加层（EffectDelta）
    EffectDelta {
        target: TargetRef,
        what: String,
        delta: i64,
    },
    /// 进入蓄力（Charge）
    Charge { target: TargetRef },
    /// 脱离/退场（Escape）
    Escape {
        target: TargetRef,
        #[serde(default)]
        inherit: bool,
        #[serde(default)]
        urgent: bool,
        #[serde(default)]
        then: Option<Vec<IrPayload>>,
    },
    /// 返场（Return）
    Return { target: TargetRef },
    /// 锁定对手换宠（Lock）
    Lock { target: TargetRef, turns: i64 },
    /// 打断目标行动（Interrupt）
    Interrupt { target: TargetRef },
    /// 交换（Exchange）
    Exchange { target: TargetRef, what: String },
    /// 重置属性（Reset）
    Reset { target: TargetRef, stat: String },
    /// 重定向下一次行动（Redirect）
    Redirect { target: TargetRef },
    /// 爆发效果授予（BurstGrant）
    BurstGrant {
        target: TargetRef,
        #[serde(default)]
        skill_where: Option<IrPayload>,
        #[serde(default)]
        skill_filter: Option<String>,
        #[serde(default)]
        effects: Vec<IrPayload>,
        #[serde(default)]
        source: String,
    },
    /// 重放已用技能（Replay）
    Replay {
        #[serde(rename = "from")]
        from_: TargetRef,
        #[serde(default)]
        skill_filter: Option<IrPayload>,
    },
    /// 借用对手当前技能（Borrow）
    Borrow { from_skill: String },
    /// 队伍计数器（TeamCounterDelta）
    TeamCounterDelta { target: String, key: String, delta: i64 },
    /// 玩家魔力/生命（LivesDelta）
    LivesDelta { target_team: String, delta: i64 },
    /// 延迟效果登记（ScheduleEntry）
    ScheduleEntry {
        turns: i64,
        #[serde(default = "default_turn_start")]
        at: String,
        #[serde(default)]
        then: Vec<IrPayload>,
    },
    /// 换宠时效果转移（InheritEffectsMutation）
    InheritEffects {
        source_key: String,
        target_key: String,
        #[serde(default = "default_battlefield")]
        scope: String,
        #[serde(default)]
        via_pending: bool,
        #[serde(default)]
        effects: Vec<IrPayload>,
        #[serde(default)]
        inherit_stat_effects: bool,
    },
    /// 变身（TransformMutation）
    Transform {
        species: String,
        #[serde(default)]
        skills: Option<Vec<String>>,
        #[serde(default)]
        reset_hp: bool,
        #[serde(default)]
        reset_energy: bool,
    },
    /// 特性压制/移除/复制（TraitInteractionMutation）
    TraitInteraction {
        action: String,
        target: String,
        #[serde(default)]
        copy_from: Option<String>,
        #[serde(default)]
        new_ability: Option<String>,
    },
    /// 授予临时技能（GainSkillsMutation）
    GainSkills {
        #[serde(default = "default_one_i")]
        count: i64,
        #[serde(default = "default_true")]
        exclude_carried: bool,
        #[serde(default = "default_learnset")]
        source: String,
        #[serde(default = "default_sprite_self")]
        target: TargetRef,
    },
    /// 注册持久计数器（CounterRegister）
    CounterRegister {
        #[serde(default)]
        name: Option<String>,
        #[serde(default)]
        cond: Option<IrPayload>,
        #[serde(default)]
        then: Vec<IrPayload>,
        #[serde(default = "default_persistent")]
        scope: String,
        #[serde(default)]
        listen: Option<Vec<String>>,
        #[serde(default = "default_one_i")]
        threshold: i64,
        #[serde(default = "default_true")]
        reset_on_fire: bool,
    },
}

pub type Journal = Vec<Mutation>;

// ── 默认值（与 Python dataclass 对齐）──
fn default_battlefield() -> String { "battlefield".into() }
fn default_set() -> String { "set".into() }
fn default_apply() -> String { "apply".into() }
fn default_one_f() -> f64 { 1.0 }
fn default_one_i() -> i64 { 1 }
fn default_steal() -> String { "steal".into() }
fn default_turn_start() -> String { "turn_start".into() }
fn default_true() -> bool { true }
fn default_learnset() -> String { "learnset".into() }
fn default_sprite_self() -> String { "sprite_self".into() }
fn default_persistent() -> String { "persistent".into() }
